beacon_node/network/src/sync/sampling.rs

```diff
@@ -0,0 +1,486 @@
+    ) -> Option<SamplingResult> {
+        let Some(request) = self.requests.get_mut(&id.id) else {
+            // TOOD(das): This log can happen if the request is error'ed early and dropped
+            debug!(self.log, "Sample downloaded event for unknown request"; "id" => ?id);
+            return None;
+        };
+
+        let result = request.on_sample_downloaded(peer_id, id.column_index, resp, cx);
+        self.handle_sampling_result(result, &id.id)
+    }
+
+    pub fn on_sample_verified(
+        &mut self,
+        id: SamplingId,
+        result: Result<(), String>,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> Option<SamplingResult> {
+        let Some(request) = self.requests.get_mut(&id.id) else {
+            // TOOD(das): This log can happen if the request is error'ed early and dropped
+            debug!(self.log, "Sample verified event for unknown request"; "id" => ?id);
+            return None;
+        };
+
+        let result = request.on_sample_verified(id.column_index, result, cx);
+        self.handle_sampling_result(result, &id.id)
+    }
+
+    fn handle_sampling_result(
+        &mut self,
+        result: Result<Option<()>, SamplingError>,
+        id: &SamplingRequester,
+    ) -> Option<SamplingResult> {
+        let result = result.transpose();
+        if result.is_some() {
+            debug!(self.log, "Removed sampling request"; "id" => ?id);
+            self.requests.remove(id);
+        }
+        result
+    }
+}
+
+pub struct ActiveSamplingRequest<T: BeaconChainTypes> {
+    block_root: Hash256,
+    block_slot: Slot,
+    requester_id: SamplingRequester,
+    column_requests: FnvHashMap<ColumnIndex, ActiveColumnSampleRequest>,
+    column_shuffle: Vec<ColumnIndex>,
+    required_successes: Vec<usize>,
+    /// Logger for the `SyncNetworkContext`.
+    pub log: slog::Logger,
+    _phantom: PhantomData<T>,
+}
+
+#[derive(Debug)]
+pub enum SamplingError {
+    SendFailed(&'static str),
+    ProcessorUnavailable,
+    TooManyFailures,
+    BadState(String),
+}
+
+/// Required success index by current failures, with p_target=5.00E-06
+/// Ref: https://colab.research.google.com/drive/18uUgT2i-m3CbzQ5TyP9XFKqTn1DImUJD#scrollTo=E82ITcgB5ATh
+const REQUIRED_SUCCESSES: [usize; 11] = [16, 20, 23, 26, 29, 32, 34, 37, 39, 42, 44];
+
+#[derive(Debug, Clone)]
+pub enum SamplingConfig {
+    Default,
+    #[allow(dead_code)]
+    Custom {
+        required_successes: Vec<usize>,
+    },
+}
+
+impl<T: BeaconChainTypes> ActiveSamplingRequest<T> {
+    fn new(
+        block_root: Hash256,
+        block_slot: Slot,
+        requester_id: SamplingRequester,
+        sampling_config: &SamplingConfig,
+        log: slog::Logger,
+    ) -> Self {
+        // Select ahead of time the full list of to-sample columns
+        let mut column_shuffle = (0..64).collect::<Vec<ColumnIndex>>();
+        let mut rng = thread_rng();
+        column_shuffle.shuffle(&mut rng);
+
+        Self {
```

Comment:
is it worth adding an extra check that `required_successes` is always less than the length of `column_shuffle`?